lib/chef/provider/user/smartos.rb

```diff
@@ -0,0 +1,48 @@
+#
+# Author:: Dave Eddy (<dave@daveeddy.com>)
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'chef/mixin/shell_out'
+require 'chef/provider/user/useradd'
+
+class Chef
+  class Provider
+    class User
+      class SmartOS < Chef::Provider::User::Useradd
+        provides :user, os: 'solaris2', platform_family: 'smartos'
+
+        def check_lock
+          shadow_line = shell_out!('getent', 'shadow', new_resource.username).stdout.strip
```

Comment:
https://github.com/chef/chef/blob/master/lib/chef/provider/user.rb#L178

```
 def action_lock
        if check_lock() == false
          converge_by("lock the user #{@new_resource.username}") do
            lock_user
            Chef::Log.info("#{@new_resource} locked")
          end
         else
          Chef::Log.debug("#{@new_resource} already locked - nothing to do")
        end
      end
```

So the code inside the converge_by() block does not get run in why-run mode, why-run would say "we would lock the user bob" but because the call to check_lock() is outside the converge_by, it would get run. The agreement is that everything that might change the system has to be called inside a converge_by.

So yeah, in this case, I think the shell_out! should get called, and we don't want it to raise.